### PR TITLE
Use BufferedInputStream for read operations

### PR DESCRIPTION
**Pull request by HaloFour**
Change Sirius read ops to use a `BufferedInputStream` over the underlying Uberstore files to improve performance, particularly around bootstrapping.

Rudimentary benchmark running XtvAPI locally on my Macbook Pro:

Sirius 1.2.6:
```
Finished firing up Session Sirius on HQSML-1685301:2553 in 501986 ms
Finished firing up Metadata Sirius on HQSML-1685301:2552 in 1237428 ms
```

Sirius 1.2.8-SNAPSHOT (with file op mods):
```
Finished firing up Session Sirius on HQSML-1685301:2553 in 187282 ms
Finished firing up Metadata Sirius on HQSML-1685301:2552 in 224490 ms
```

### CODE REVIEW

**Comment by HaloFour** [comment]
Testing on a prod-sized XtvAPI instance with full metadata and session Uberstores:

Sirius 1.2.6:
```
Session BootstrapTime  : Some(Some(296165)) - 4.9 minutes
Metadata BootstrapTime  : Some(Some(2054482)) - 34 minutes
```

Sirius 1.2.8-SNAPSHOT (with buffering enabled, 8k buffer):
```
Session BootstrapTime  : Some(Some(166132)) - 2.8 minutes
Metadata BootstrapTime  : Some(Some(1375710)) - 22.9 minutes
```

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFile.scala [hunk 1]
```diff
@@ -50,7 +79,9 @@ object UberDataFile {
    *
    * @param dataFileName the name of the file for this descriptor to wrap
    */
-  private[data] class UberFileDesc(dataFileName: String) {
+  private[data] class UberFileDesc(dataFileName: String,
+                                   useReadBuffer: Boolean,
+                                   readBufferSize: Int) {
```
Comment (clinedome) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFile.scala]: I think the `UberFileDesc` is actually the place to differentiate between the types of `UberDataFileReader` here. The `createReadHandle` can use the buffer type and choose whether to make a `BufferedUberDataFileReader` or a `RandomAccessUberDataFileReader`. Holding the config values just to pass them into an `apply` later doesn't seem to make lots of sense, and this class is all about generating these things anyway.

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    val byte3 = this.read()
+    val byte4 = this.read()
+    if (byte2 < 0 || byte3 < 0 || byte4 < 0) {
+      throw new EOFException()
+    }
+    Some((byte1 << 24) + (byte2 << 16) + (byte3 << 8) + (byte4 << 0))
+  }
+}
+
+object UberDataFileReader {
```
Comment (clinedome) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala]: If you move the logic for reading the `useReadBuffer` to the `UberFileDesc` class, this case object can just go away, and `sequentialBuffered` and `randomAccess` can become `apply` methods on the `BufferedUberDataFileReader` and `RandomAccessUberDataFileReader` objects, respectively.

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  def randomAccess(dataFileName: String, baseOffset: Long): UberDataFileReader = {
+    val randomAccessFile = new RandomAccessFile(dataFileName, "r")
+    randomAccessFile.seek(baseOffset)
+    new RandomAccessUberDataFileReader(randomAccessFile)
+  }
+}
+
+private[uberstore] class RandomAccessUberDataFileReader(val randomAccessFile: RandomAccessFile) extends UberDataFileReader {
```
Comment (clinedome) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala]: I think it'd be worth moving `RandomAccessUberDataFileReader` and `BufferedUberDataFileReader` to separate files at this point.

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+      return None
+    }
+    val byte2 = this.read()
+    val byte3 = this.read()
+    val byte4 = this.read()
+    if (byte2 < 0 || byte3 < 0 || byte4 < 0) {
+      throw new EOFException()
+    }
+    Some((byte1 << 24) + (byte2 << 16) + (byte3 << 8) + (byte4 << 0))
+  }
```
Comment (clinedome) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala]: So this sort of a _read if not eof_ kind of thing. I don't love having to do the bitwise operations here just because we're using a read as a test. This approach distinguishes between _there are no bytes to read_ and _there were 1-3 bytes to read_. I'm not sure we care too much about that. I wonder if we should just wrap `readInt` in a `Result` and convert it to an `Option` on the way out.

The behavior is also a little confusing, at least as stated. I'd suggest either making it very explicit that this is a `readIntIfNotEof` or throwing those details in the comments. Generally something called `tryReadInt` I would not expect to explicitly throw an exception.

The alternative is to build an explicit `eof(): Boolean` function into each of the implementations for `UberDataFile` readers.

Thoughts?

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala [hunk 5]
```diff
@@ -286,16 +291,16 @@ class SegmentedUberStore private[segmented] (base: JFile, eventsPerSegment: Long
    */
   private def replaceSegments(replacementMap: Map[Segment, String]) = replaceLock.synchronized {
     for (segment <- replacementMap.keys) {
-      replaceSegment(segment, replacementMap(segment))
+      replacesegment(segment, replacementMap(segment))
```
Comment (clinedome) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala]: I think the `find/replace` got carried away on these, don't need to lower-case the `S` in segment here and in the two below

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactor.scala [hunk 6]
```diff
@@ -33,12 +33,14 @@ object SegmentedCompactor {
...
-    new SegmentedCompactor(maxDeleteAgeMillis)
+    new SegmentedCompactor(maxDeleteAgeMillis, siriusConfig)
   }
 }
 
-private [segmented] class SegmentedCompactor(maxDeleteAgeMillis: Long) {
+private [segmented] class SegmentedCompactor(maxDeleteAgeMillis: Long, siriusConfig: SiriusConfiguration) {
 
+  def segment(base: JFile, name: String): Segment = Segment(siriusConfig, base, name)
+  def segment(location: JFile): Segment = Segment(siriusConfig, location)
```
Comment (clinedome) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactor.scala]: I don't love having to weave the config into the class instantiation here and pass it through to our `Segment` instantiation.

The alternative would be to have apply pass a `buildSegment: (JFile => Segment)` as a parameter to the `SegmentedCompactor` class. I might actually prefer that tbh, as it keeps all the configuration mess in the `apply` function.

Like:
```
  def apply(siriusConfig: SiriusConfiguration) : SegmentedCompactor = {
    def buildSegment(file: JFile): Segment = Segment(siriusConfig, file)

    val maxDeleteAgeHours = siriusConfig.getInt(SiriusConfiguration.COMPACTION_MAX_DELETE_AGE_HOURS, Integer.MAX_VALUE)
    val maxDeleteAgeMillis = 60L * 60 * 1000 * maxDeleteAgeHours

    new SegmentedCompactor(maxDeleteAgeMillis, buildSegment)
  }
```

```
private [segmented] class SegmentedCompactor(maxDeleteAgeMillis: Long, buildSegment: (JFile => Segment)) {
  def buildSegment(base: JFile, name: String): Segment = buildSegment(new JFile(base, name))
  // ...
```

Also I'd generally prefer `buildSegment` to `segment` as far as naming is concerned, just to make it a little clearer what's happening.

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala (hunk 4 above)
Comment (HaloFour) [reply src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala]: IMO I think it's worthwhile to distinguish between actually being at the end of the file vs. having a partial read as the latter would indicate a corrupt Uberstore.

I looked into having a separate `eof()` method but the issue with that was the lack of a good way of peeking into the contents of the stream without also advancing the stream.  The alternative would be to constantly check the size of the file which appeared to add a lot of overhead (I don't have that number on hand).  I can run a benchmark on setting a mark on the stream and see what that does.

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFile.scala [hunk 7]
```diff
@@ -16,12 +16,34 @@
...
+    *
+    * @param dataFileName the data file name, this will
+    *          be created if it does not exist
+    *
+    * @return fully constructed UberDataFile
+    */
+  def apply(dataFileName: String): UberDataFile = {
+
+    apply(dataFileName, useReadBuffer = false, readBufferSize = 0)
+  }
```
[the remark this answers is not in the document]
Comment (HaloFour) [reply src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFile.scala]: I'll look into refactoring it out.  If I recall the issue I ran into was that this method was being invoked from a path where a `SiriusConfiguration` was not available, such as from `WalTool`

[the remark this answers is not in the document]
Comment (HaloFour) [reply src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFile.scala]: I've refactored the passing of `SiriusConfiguration` back to `UberStore` at which point I made it a default parameter.  `UberStore.apply` is used in a number of locations where it didn't make sense to pass a `SiriusConfiguration`, such as in `WalTool`.  If you'd prefer I could make it a required parameter there as well and just pass empty `SiriusConfiguration` instances, but that seemed a bit unnecessary to me.

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala (hunk 4 above)
Comment (HaloFour) [reply src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala]: The mark/reset approach on the `InputStream` worked fairly well so I've moved to using an `eof` method.  Raw I/O ops were _slightly_ slower in a synthetic benchmark but bootstrapping a full XtvAPI didn't take any longer.

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/data/BufferedUberDataFileReader.scala [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+  override def read(b: Array[Byte], off: Int, len: Int): Int = channel.read(ByteBuffer.wrap(b, off, len))
+
+  override def skip(n: Long): Long = {
+    val newPosition = channel.position() + n
+    channel.position(newPosition)
+    newPosition
+  }
+
+  override def close(): Unit = channel.close()
+}
```
Comment (clinedome) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/data/BufferedUberDataFileReader.scala]: newline

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    */
+  def readLong(): Long
+
+  /**
+    * Populates the entire array with the next bytes from the file
+    * @param array to populate
+    * @throws EOFException if EOF
+    */
+  def readFully(array: Array[Byte]): Unit
+}
```
Comment (clinedome) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala]: newline

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala [hunk 10]
```diff
@@ -257,7 +262,7 @@ class SegmentedUberStore private[segmented] (base: JFile, eventsPerSegment: Long
         segmentedCompactor.mergeSegments(left, right, merged)
         replaceLock synchronized {
           replaceSegment(left, merged.getAbsolutePath)
-          removeSegment(right)
+          removesegment(right)
```
Comment (clinedome) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala]: can undo this change

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala [hunk 11]
```diff
@@ -97,7 +97,10 @@ object SegmentedUberStore {
  *
  * @param base directory SegmentedUberStore is based in
  */
-class SegmentedUberStore private[segmented] (base: JFile, eventsPerSegment: Long, segmentedCompactor: SegmentedCompactor) extends SiriusLog {
+class SegmentedUberStore private[segmented] (base: JFile,
+                                             eventsPerSegment: Long,
+                                             segmentedCompactor: SegmentedCompactor,
+                                             siriusConfig: SiriusConfiguration) extends SiriusLog {
```
Comment (clinedome) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala]: I'd suggest we use the same pattern of making the `buildSegment` in the `apply` function, so we're not passing the configuration into the class itself. I much prefer passing the exact parameters required by the class, and letting the apply method be the thing that extracts whatever's necessary from the config in order to build those params.

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala [hunk 12]
```diff
@@ -295,7 +300,7 @@ class SegmentedUberStore private[segmented] (base: JFile, eventsPerSegment: Long
     readOnlyDirs = replaceElement(readOnlyDirs, original, newSegment)
   }
 
-  private def removeSegment(segment: Segment) = replaceLock.synchronized {
+  private def removesegment(segment: Segment) = replaceLock.synchronized {
```
Comment (clinedome) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala]: can undo this change

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/Segment.scala [hunk 13]
```diff
@@ -42,15 +44,15 @@ object Segment {
...
+  def apply(siriusConfig: SiriusConfiguration, location: File): Segment = {
     location.mkdirs()
 
     val dataFile = new File(location, "data")
     val indexFile = new File(location, "index")
     val compactionFlagFile = new File(location, "keys-collected")
     val internalCompactionFlagFile = new File(location, "internally-compacted")
 
-    val data = UberDataFile(dataFile.getAbsolutePath)
+    val data = UberDataFile(dataFile.getAbsolutePath, siriusConfig)
```
Comment (derjust) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/Segment.scala]: Maybe let `UberFileDesc` take carer off the config & produce the actual file object?
Feels a little strange having the config just for another layer to pass through.

Comment (HaloFour) [reply src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/Segment.scala]: That's not a bad idea.

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+  def tryReadInt(): Option[Int] = {
+    val byte1 = this.read()
+    if (byte1 < 0) {
+      return None
+    }
+    val byte2 = this.read()
+    val byte3 = this.read()
+    val byte4 = this.read()
+    if (byte2 < 0 || byte3 < 0 || byte4 < 0) {
+      throw new EOFException()
```
[the remark this answers is not in the document]
Comment (HaloFour) [reply src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala]: This method has since been refactored out.  I was looking for a cheap way of detecting EOF but separately from a partial read as the latter indicates a corrupt Uberstore.  I've found that using `mark`/`reset` of a single byte to peek into the stream works well enough and have used that to add an `eof()` method instead.

File: src/main/scala/com/comcast/xfinity/sirius/api/SiriusConfiguration.scala [hunk 15]
```diff
@@ -188,6 +188,17 @@ object SiriusConfiguration {
...
+  /**
+    * Whether or not to use a buffered reader when reading through the log
+    */
+  final val LOG_USE_READ_BUFFER = "sirius.log.read-buffer-enabled"
+
+  /**
+    * Indicates the size of the buffer in bytes when reading events from the log.
+    * The default buffer size is 8192 bytes. (int)
+    */
+  final val LOG_READ_BUFFER_SIZE = "sirius.log.read-buffer-size"
```
Comment (clinedome-work) [inline src/main/scala/com/comcast/xfinity/sirius/api/SiriusConfiguration.scala]: Addendum nitpick: we should indicate the units in the prop name/values. like `final val LOG_READ_BUFFER_SIZE_BYTES = "sirius.log.read-buffer-size-bytes"`

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala [hunk 16]
```diff
@@ -85,9 +85,10 @@ object SegmentedUberStore {
 
     repair(base)
 
+    def buildSegment(base: JFile, name: String) = Segment(siriusConfig, base, name)
     val segmentedCompactor = SegmentedCompactor(siriusConfig)
 
-    new SegmentedUberStore(new JFile(base), MAX_EVENTS_PER_SEGMENT, segmentedCompactor)
+    new SegmentedUberStore(new JFile(base), MAX_EVENTS_PER_SEGMENT, segmentedCompactor, buildSegment)
```
Comment (clinedome) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala]: Ok, so here's a thought. We've been trying to get the config weaved down nicely to the bottom layers so we can build the right read/write handles, according to that config. Instead of going to all that trouble, how about we just build an `UberFileDesc` (maybe with a rename??) _once_ here... or even up in SiriusFactory... based on the config. The whole job of that thing is to build the right read/write handles, and to do it on demand. Building one of those that is used by all the different segments makes sense I think.

We can pass that down wherever it needs to go. Segments can just accept that as a parameter, since it's all they're doing anyway, and we can get rid of some of these config-hiding abstractions like passing in `buildSegment` it also limits the scope of the config creep.

Comment (HaloFour) [reply src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala]: So instead of each instance of `UberFileDesc` (name pending) existing per data file there'd be one shared instance that can open a handle to any file?

Comment (clinedome) [reply src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala]: Yeah, something along those lines. All it does is take file paths and build the right kinds of input streams / output file handles, based on configured behavior. None of the lower-level pieces (Segment, etc) care about _how_ it does its job. They just need something that can get them a `.foreach` from a file. Passing in something from above that can do that translation (sort of analogous to the way we do the event codec) makes sense imo.

Comment (HaloFour) [reply src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala]: ok, I've pushed a commit which includes some of the refactoring.  I went with the name `UberStoreFileHandleFactory` and made it a trait.  I've not yet touched the unit tests and it needs scaladocs but I wanted to get input on the refactoring first.

Comment (HaloFour) [reply src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala]: I've refactored a bit more around.  I changed some of the names around to be (what I think is) more consistent.  So the new trait is `UberDataFileHandleFactory` and it can create `UberDataFileReadHandle` and `UberDataFileWriteHandle` instances.  I still pass the `buildSegment` function around but I could easily refactor that away, although I like the idea that it's kept abstracted away and only `SegmentedUberStore` knows how its implemented.

I did update the existing unit tests this time and they all pass but this would require some additional tests, plus scaladocs, but again I'd like some feedback as to the direction.

Comment (derjust) [reply src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala]: I like it & I have no other comments on how it is right now beyond what you lined out already 👍

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala [hunk 17]
```diff
@@ -295,7 +304,7 @@ class SegmentedUberStore private[segmented] (base: JFile, eventsPerSegment: Long
     readOnlyDirs = replaceElement(readOnlyDirs, original, newSegment)
   }
 
-  private def removeSegment(segment: Segment) = replaceLock.synchronized {
+  private def replaceSegment(segment: Segment) = replaceLock.synchronized {
```
Comment (clinedome) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala]: This should probably still be `removeSegment`, right? We're not doing anything other than deleting it.

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/data/BufferedFileReadHandle.scala [hunk 18]
```diff
@@ -0,0 +1,114 @@
+/*
+ *  Copyright 2012-2014 Comcast Cable Communications Management, LLC
```
[the remark this answers is not in the document]
Comment (HaloFour) [reply src/main/scala/com/comcast/xfinity/sirius/uberstore/data/BufferedFileReadHandle.scala]: I was going to ask about that.  Should we bump all of the copyright headers?

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala (hunk 17 above)
Comment (HaloFour) [reply src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedUberStore.scala]: Dang it, IntelliJ...

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/data/BufferedFileReadHandle.scala (hunk 18 above)
[the remark this answers is not in the document]
Comment (clinedome) [reply src/main/scala/com/comcast/xfinity/sirius/uberstore/data/BufferedFileReadHandle.scala]: Yeah, maybe let's do that in a separate PR
